### PR TITLE
ADBDEV-4203: Refactor closing connection to pxf, including fdw

### PR DESCRIPTION
**Pull request by RekGRpth**
The [patch](https://github.com/arenadata/pxf/pull/42) does not cover the fdw.
Also, that patch only works on the top-level transaction, while on sub-transactions it would be better not to wait so long.
Also, some resources in that patch are allocated in different contexts, and it may happen that they are freed before or after the callback, which can lead to a segfault.
More criticism of the old patch can be found in the [upstream](https://github.com/greenplum-db/pxf/pull/992).

The C-part of the PXF external table releases the context (`cleanup_context`) only on the last call in the `pxfprotocol_export` and `pxfprotocol_import` functions.
The C-part of the PXF fdw releases the context (`PxfBridgeCleanup`) only in the `FinishForeignModify` function.
That is, on errors, the context is not released, including curl-connections to the Java-part are not closed.
Therefore, I added a callback to release resources on errors, which releases the context, including closing curl-connections.

For example, this may happen when a request is canceled when an exclusive lock has been taken on a table referenced by an external table or fdw.
1) create external table
```sql
CREATE EXTENSION pxf;
CREATE TABLE t(a int) DISTRIBUTED BY (a);
INSERT INTO t SELECT i FROM generate_series(1, 1000) i;
CREATE EXTERNAL TABLE e(a int) LOCATION('pxf://t?PROFILE=JDBC&JDBC_DRIVER=org.postgresql.Driver&DB_URL=jdbc:postgresql://localhost:6432/postgres') FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');
```
or create fdw
```sql
CREATE EXTENSION pxf_fdw;
CREATE TABLE t(a int) DISTRIBUTED BY (a);
INSERT INTO t SELECT i FROM generate_series(1, 1000) i;
CREATE SERVER s FOREIGN DATA WRAPPER jdbc_pxf_fdw OPTIONS (jdbc_driver 'org.postgresql.Driver', db_url 'jdbc:postgresql://localhost:6432/postgres');
CREATE USER MAPPING FOR CURRENT_USER SERVER s;
CREATE FOREIGN TABLE e (a int) SERVER s OPTIONS (resource 't');
```
2) in one session, run the command:
```sql
begin;lock table t in ACCESS exclusive mode;
```
and don't close the session.

3) in another session, run the command:
```sql
SELECT count(*) FROM e;
```
and then interrupt its execution with Ctrl+C, but don't close the session.
The connection from the C-part to the Java-part remains:
```sh
bash-4.2$ netstat -enpt | grep :5888
tcp        0      0 127.0.0.1:42236         127.0.0.1:5888          ESTABLISHED 1000       7244695    161781/postgres:  6 
tcp        0      0 127.0.0.1:5888          127.0.0.1:42236         ESTABLISHED 1000       7229437    157250/java         
```
expected (and with patch): adb connection is closed, but Java-part remain:
```sh
bash-4.2$ netstat -enpt | grep :5888
tcp        1      0 127.0.0.1:5888          127.0.0.1:47074         CLOSE_WAIT  1000       6885317    144206/java         
```
Another example with subtransaction
1) create external table
```sql
CREATE EXTENSION pxf;
CREATE TABLE t(a int) DISTRIBUTED BY (a);
INSERT INTO t SELECT i FROM generate_series(1, 1000) i;
CREATE EXTERNAL TABLE e(a int, b int) LOCATION('pxf://t?PROFILE=JDBC&JDBC_DRIVER=org.postgresql.Driver&DB_URL=jdbc:postgresql://localhost:6432/postgres') FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');
```
or create fdw
```sql
CREATE EXTENSION pxf_fdw;
CREATE TABLE t(a int) DISTRIBUTED BY (a);
INSERT INTO t SELECT i FROM generate_series(1, 1000) i;
CREATE SERVER s FOREIGN DATA WRAPPER jdbc_pxf_fdw OPTIONS (jdbc_driver 'org.postgresql.Driver', db_url 'jdbc:postgresql://localhost:6432/postgres');
CREATE USER MAPPING FOR CURRENT_USER SERVER s;
CREATE FOREIGN TABLE e (a int, b int) SERVER s OPTIONS (resource 't');
```
2) run commands:
```sql
BEGIN;
INSERT INTO t SELECT i FROM generate_series(1, 1000) i;
SAVEPOINT a;
SELECT count(*) FROM e;
```
and don't close the session.
The connection from the C-part to the Java-part remains:
```sh
bash-4.2$ netstat -enpt | grep :5888
tcp        6      0 127.0.0.1:35230         127.0.0.1:5888          CLOSE_WAIT  1000       2135014    36419/postgres:  64 
```
expected (and with patch): adb connection is closed.

### CODE REVIEW

**Comment by InnerLife0** [comment]
Any single word about why previous solution not worked and how current solution differs from previous one?

**Comment by RekGRpth** [comment]
> Any single word about why previous solution not worked and how current solution differs from previous one?

See in tracker

**Comment by RekGRpth** [comment]
> > Any single word about why previous solution not worked and how current solution differs from previous one?
> 
> See in tracker

I added it into PR description

**Comment by InnerLife0** [comment]
> More criticism of the old patch can be found in the

Give me the FULL list of problems from previous solution and the full description of solution YOU chose to resolve them. We have the second PR with near one-to-one description.

I can't look into your brain (~~I mean I can, but only once~~). Why do you think that everyone who looks to this PR, or to this commit in the future, should gather all the reasons from multiple sources (including private one), summarize them, keeping in mind that one problem was already described here, another described here(e.g. twice), or etc?

**Comment by RekGRpth** [comment]
> FULL list of problems from previous solution

1) does not cover fwd
2) works only on top-level transaction
3) callback may use resources which are pfreed already

**Comment by RekGRpth** [comment]
> full description of solution YOU chose to resolve them

1) I copied and tested solution to fwd
2) I changed the callback to work with any transaction levels
3) I removed unnecessary allocations and did not use resources from another memory context in the callback

**Comment by InnerLife0** [comment]
> * does not cover fwd
> * works only on top-level transaction
> * callback may use resources which are pfreed already

> * I copied and tested solution to fwd
> * I changed the callback to work with any transaction levels
> * I removed unnecessary allocations and did not use resources from another memory context in the callback

What about the underwater rocks with [headers](https://github.com/greenplum-db/pxf/pull/992#issuecomment-1680608725)? Why don't you highlight the importance of current solution for them?

**Comment by RekGRpth** [comment]
> Why don't you highlight the importance of current solution for them?

Headers were also freed in the previous solution.

**Comment by InnerLife0** [comment]
> > Why don't you highlight the importance of current solution for them?
> 
> Headers were also freed in the previous solution.

So there is nothing in current solution which you need to highlight about the headers?

**Comment by RekGRpth** [comment]
> So there is nothing in current solution which you need to highlight about the headers?

I free headers by another way

**Comment by InnerLife0** [comment]
> > So there is nothing in current solution which you need to highlight about the headers?
> 
> I free headers by another way

By what way? What is the difference with previous solution. Why do you ever need it?

<details><summary></summary>
Did I ever tell you what the definition of insanity is? Insanity is doing the exact... same fucking thing... over and over again expecting... shit to change... That. Is. Crazy. 
image
</details>

**Comment by RekGRpth** [comment]
> By what way? What is the difference with previous solution. Why do you ever need it?

In previous solution headers were in hadoop context, which may be pfreed before calling callback.
I just copied curl pointer of headers to context, allocated in top memory context, to access it in callback.

**Comment by RekGRpth** [comment]
I suggest to review only second commit of this PR

File: external-table/src/libchurl.c [hunk 1]
```diff
@@ -525,18 +534,56 @@ churl_cleanup(CHURL_HANDLE handle, bool after_error)
...
+
+			churl_cleanup(curr, !isCommit);
+		}
+	}
+}
+
+static churl_context *
+churl_new_context(void)
+{
+	churl_context *context = MemoryContextAllocZero(TopMemoryContext, sizeof(churl_context));
```
Comment (red1452) [inline external-table/src/libchurl.c:572]: Do we really need to use `TopMemoryContext` here?

File: external-table/src/libchurl.c [hunk 2]
```diff
@@ -736,36 +783,22 @@ multi_remove_handle(churl_context *context)
...
-			if (context->download_buffer->ptr)
-				pfree(context->download_buffer->ptr);
-			pfree(context->download_buffer);
-		}
-		if (context->upload_buffer)
-		{
-			if (context->upload_buffer->ptr)
-				pfree(context->upload_buffer->ptr);
-			pfree(context->upload_buffer);
-		}
```
Comment (red1452) [inline external-table/src/libchurl.c:768]: If these buffers were allocated at `TopMemoryContext`, It will be freed only when process will finish. Do we need to keep it all time or can we free it earlier?

File: external-table/src/libchurl.c [hunk 3]
```diff
@@ -736,36 +783,22 @@ multi_remove_handle(churl_context *context)
...
+		if (context->prev)
+			context->prev->next = context->next;
+		else
+			open_curl_handles = open_curl_handles->next;
+		if (context->next)
+			context->next->prev = context->prev;
+
+		/* memory is freed by deleting memory context */
+		context->download_buffer.ptr = NULL;
+		context->upload_buffer.ptr = NULL;
```
Comment (red1452) [inline external-table/src/libchurl.c]: I think it is unnecessary actions, because below, you are freeing all context.

File: fdw/libchurl.c [hunk 4]
```diff
@@ -42,7 +43,7 @@
 /*
  * internal buffer for libchurl internal context
  */
-typedef struct churl_buffer
+typedef struct
```
Comment (red1452) [inline fdw/libchurl.c]: To minimize diff with upstream you should remove this cosmetic fix.

File: fdw/libchurl.c [hunk 5]
```diff
@@ -51,10 +52,18 @@ typedef struct churl_buffer
...
+{
+	struct curl_slist *headers;
+} churl_settings;
+
 /*
  * internal context of libchurl
  */
-typedef struct churl_handle
+typedef struct churl_context
```
Comment (red1452) [inline fdw/libchurl.c:66]: And this.

File: external-table/src/libchurl.c [hunk 6]
```diff
@@ -525,18 +534,56 @@ churl_cleanup(CHURL_HANDLE handle, bool after_error)
...
+	}
+}
+
+static churl_context *
+churl_new_context(void)
+{
+	churl_context *context = MemoryContextAllocZero(TopMemoryContext, sizeof(churl_context));
+
+	context->owner = CurrentResourceOwner;
+	context->prev = NULL;
```
Comment (red1452) [inline external-table/src/libchurl.c]: This pointer was allocated by `MemoryContextAllocZero`. Setting NULL is unnecessary.

File: external-table/src/libchurl.c [hunk 7]
```diff
@@ -525,18 +534,56 @@ churl_cleanup(CHURL_HANDLE handle, bool after_error)
...
+	context->next = open_curl_handles;
+	if (open_curl_handles)
+		open_curl_handles->prev = context;
```
Comment (red1452) [inline external-table/src/libchurl.c:579]:
```c
	if (open_curl_handles)	
	{
		context->next = open_curl_handles;	
		open_curl_handles->prev = context;
	}
```

File: external-table/src/libchurl.c (hunk 1 above)
Comment (RekGRpth) [reply external-table/src/libchurl.c:572]: Yes, because if there are no errors, the current context will be [deleted](https://github.com/arenadata/gpdb/blob/4fa8a09d5babb471a36b8a06b499247d24d1f6ad/src/backend/access/external/url_custom.c#L115-L120) before the callback is called.

File: external-table/src/libchurl.c (hunk 2 above)
Comment (RekGRpth) [reply external-table/src/libchurl.c:768]: > If these buffers were allocated at `TopMemoryContext`

No! These buffers are [allocated in the current context](https://github.com/arenadata/pxf/blob/561c3e1bee78d4b64def5202257d6e1f685c96a5/external-table/src/libchurl.c#L1338)!

File: external-table/src/libchurl.c (hunk 3 above)
Comment (RekGRpth) [reply external-table/src/libchurl.c]: Exactly! Removed

File: external-table/src/libchurl.c (hunk 6 above)
Comment (RekGRpth) [reply external-table/src/libchurl.c]: Exactly! Removed

File: external-table/src/libchurl.c (hunk 7 above)
Comment (RekGRpth) [reply external-table/src/libchurl.c:579]: applied

File: fdw/libchurl.c (hunk 4 above)
Comment (RekGRpth) [reply fdw/libchurl.c]: done

File: fdw/libchurl.c (hunk 5 above)
Comment (RekGRpth) [reply fdw/libchurl.c:66]: I can't do that here because it's used [below](https://github.com/arenadata/pxf/pull/53/files#diff-c3762d4b3b54cb8b260cd18189402d14880c8641c933af06896c931821996719R104-R105).

**Review by red1452 — APPROVED** [review]

File: fdw/libchurl.c [hunk 8]
```diff
@@ -746,36 +794,18 @@ multi_remove_handle(churl_context *context)
...
-}
-
 static void
 churl_cleanup_context(churl_context *context)
 {
 	if (context)
 	{
-		if (context->download_buffer)
-		{
-			if (context->download_buffer->ptr)
```
Comment (InnerLife0) [inline fdw/libchurl.c:769]: Why did you remove explicit buffer free? Is default mem context always freed after churl_cleanup_context() call?

File: external-table/src/pxfbridge.c [hunk 9]
```diff
@@ -40,11 +40,7 @@ gpbridge_cleanup(gphadoop_context *context)
 	if (context == NULL)
 		return;
 
-	churl_cleanup(context->churl_handle, context->after_error);
-	context->churl_handle = NULL;
```
Comment (InnerLife0) [inline external-table/src/pxfbridge.c:44]: Why did we remove this line?

**Comment by InnerLife0** [comment]
The current version of patch covers more problems comparing to the previous version.
The reproduction case you mentioned covers only connection closing.
How can I test that other code works well? Is it possible to automate this testing?

Was it approved by team leader that we need to fix fdw and external-table parts all at once?

File: fdw/libchurl.c (hunk 8 above)
Comment (RekGRpth) [reply fdw/libchurl.c:769]: > Why did you remove explicit buffer free?

Because now the churl_cleanup_context function can be called both before and after the context in which the buffers were allocated is freed.

Comment (RekGRpth) [reply fdw/libchurl.c:769]: > Is default mem context always freed after churl_cleanup_context() call?

[Yes](https://github.com/arenadata/gpdb/blob/4fa8a09d5babb471a36b8a06b499247d24d1f6ad/src/backend/access/external/url_custom.c#L115-L120)

File: external-table/src/pxfbridge.c (hunk 9 above)
Comment (RekGRpth) [reply external-table/src/pxfbridge.c:44]: > Why did we remove this line?

Because nowhere is it checked whether it is null or not.

**Comment by RekGRpth** [comment]
> Was it approved by team leader that we need to fix fdw and external-table parts all at once?

Yes, why not? Moreover, the code in them is almost the same.

**Comment by RekGRpth** [comment]
> How can I test that other code works well?

I added example with subtransaction to PR description.

**Comment by RekGRpth** [comment]
> The reproduction case you mentioned covers only connection closing.

But this patch is just about closing the connection! Or did I misunderstand your statement?

**Comment by RekGRpth** [comment]
> Is it possible to automate this testing?

I think that, in principle, it is possible, but then PXF needs an infrastructure such as isolation tests so that you can make multiple connections and execute shell commands.

**Comment by InnerLife0** [comment]
> Yes, why not? Moreover, the code in them is almost the same.

My general idea and the main question to any of patch (your or not your) is how to check that changes not break something existed. Reviewing something separated is easier, like step by step.
About this patch: It seems it's OK for the team and I finished with debugging and manual review of both parts. Now it doesn't matter. Thanks.

> But this patch is just about closing the connection! Or did I misunderstand your statement?

Properly closed connection is just a symptom which is visible to end user. Under the hood your patch now also contains a proper work with context and all surrounding changes and optimizations which are poorly commented.

> I think that, in principle, it is possible, but then PXF needs an infrastructure such as isolation tests so that you can make multiple connections and execute shell commands.

Looking to manual test example, I see that multiple connections are not necessary. Regression tests engine is already available for external-table.
But anyway, I think I'm done here.

**Review by InnerLife0 — APPROVED** [review]
